shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm

```diff
@@ -0,0 +1,193 @@
+
+  // The context provided to the Flutter engine for rendering to the FlutterView. This is lazily
+  // created during initialization of the FlutterView. This is used to render content into the
+  // FlutterView.
+  NSOpenGLContext* _openGLContext;
+
+  // The context provided to the Flutter engine for resource loading.
+  NSOpenGLContext* _resourceContext;
+
+  // A mapping of textureID to internal FlutterExternalTextureGL adapter.
+  NSMutableDictionary<NSNumber*, FlutterExternalTextureGL*>* _textures;
+
+  FlutterEngine* _flutterEngine;
+}
+
+- (instancetype)initWithEmbedderEngine:(FLUTTER_API_SYMBOL(FlutterEngine))engine
+                         flutterEngine:(FlutterEngine*)flutterEngine {
+  self = [super init];
+  if (self) {
+    _engine = engine;
+    _flutterEngine = flutterEngine;
+    _textures = [[NSMutableDictionary alloc] init];
+  }
+  return self;
+}
+
+- (void)attachToFlutterView:(FlutterView*)view {
+  _flutterView = view;
+}
+
+- (bool)makeCurrent {
+  if (!_openGLContext) {
+    return false;
+  }
+  [_openGLContext makeCurrentContext];
+  return true;
+}
+
+- (bool)clearCurrent {
+  [NSOpenGLContext clearCurrentContext];
+  return true;
+}
+
+- (bool)present {
+  if (!_openGLContext) {
+    return false;
+  }
+  [_flutterView present];
+  return true;
+}
+
+- (uint32_t)fboForFrameInfo:(const FlutterFrameInfo*)info {
+  CGSize size = CGSizeMake(info->size.width, info->size.height);
+  return [_flutterView frameBufferIDForSize:size];
+}
+
+- (NSOpenGLContext*)resourceContext {
+  if (!_resourceContext) {
+    NSOpenGLPixelFormatAttribute attributes[] = {
+        NSOpenGLPFAColorSize, 24, NSOpenGLPFAAlphaSize, 8, 0,
+    };
+    NSOpenGLPixelFormat* pixelFormat = [[NSOpenGLPixelFormat alloc] initWithAttributes:attributes];
+    _resourceContext = [[NSOpenGLContext alloc] initWithFormat:pixelFormat shareContext:nil];
+  }
+  return _resourceContext;
+}
+
+- (NSOpenGLContext*)openGLContext {
+  if (!_openGLContext) {
+    NSOpenGLContext* shareContext = [self resourceContext];
+    _openGLContext = [[NSOpenGLContext alloc] initWithFormat:shareContext.pixelFormat
+                                                shareContext:shareContext];
+  }
+  return _openGLContext;
+}
+
+- (bool)makeResourceCurrent {
+  [self.resourceContext makeCurrentContext];
+  return true;
+}
+
+- (void)clearResourceContext {
+  _resourceContext = nil;
+}
+
+#pragma mark - FlutterTextureRegistrar
+
+- (BOOL)populateTextureWithIdentifier:(int64_t)textureID
+                        openGLTexture:(FlutterOpenGLTexture*)openGLTexture {
+  return [_textures[@(textureID)] populateTexture:openGLTexture];
+}
+
+- (int64_t)registerTexture:(id<FlutterTexture>)texture {
+  FlutterExternalTextureGL* FlutterTexture =
+      [[FlutterExternalTextureGL alloc] initWithFlutterTexture:texture];
+  int64_t textureID = [FlutterTexture textureID];
+  FlutterEngineResult success =
+      _flutterEngine.embedderAPI.RegisterExternalTexture(_engine, textureID);
```

Comment:
These calls should be wrapped as private FlutterEngine methods, rather than having other classes calling into embedderAPI directly.